### ISSUE TITLE
Multilingual ASR with Auxiliary CTC objectives

### ISSUE THREAD

**Issue by david-gimeno** [espnet/espnet#5774]
In the paper entitled: "Improving Massively Multilingual ASR with Auxiliary CTC Objectives" you proposed a method where early encoder layers where conditioned to identify the language of speech, while the remaining ones followed the conventional InterCTC strategy. My question is:

is there a already implemented version of this model architecture in ESPNet? Or do I have to implemented on myself?

Thanks in advance,

David.

**Comment by sw005320** [comment]
Yes, you can refer to https://github.com/espnet/espnet/tree/master/egs2/fleurs/asr1#multilingual-asr---ssl--e-branchformer--self-condition-xls-r-e-branchformer-utt_mvn-6500-bpe
I added @wanchichen in this thread.

**Comment by david-gimeno** [comment]
When I click the link to see the config yaml file, a 404 not found raises. Usually I am using ESPNet at python level. My question is: where can I find that Conformer-LID encoder version?

Thanks :)

**Comment by sw005320** [comment]
@wanchichen, can you answer it?
Also, please fix the link

**Comment by wanchichen** [comment]
Its in https://github.com/espnet/espnet/blob/master/egs2/fleurs/asr1/conf/tuning/train_asr_conformer_hier_lid_utt.yaml

I'll fix the link, it was missing `/tuning/`.

FYI the batch size in this recipe is not very efficient, you can probably increase it and decrease the `accum_grad`.

**Comment by david-gimeno** [comment]
Thank you so much :) Everything is clear now. Have a nice day!

**Comment by david-gimeno** [comment]
I have some questions about the _Hierarchical Utternace-level LID_ framework because I do not manage to make the model work properly.  Thanks to the [configuration](https://github.com/espnet/espnet/blob/master/egs2/fleurs/asr1/conf/tuning/train_asr_conformer_hier_lid_utt.yaml)  you shared to me some weeks ago, I learn how to use and include auxiliary CTC tasks. Specifically, I built a 12-layer Conformer encoder followed by a MaskCTC decoder to tackle two different languages. I am using InterCTC losses in the corresponding 3, 6, and 9 layers. 

However, the model is not learning. Conversely, if I am training a Conformer-based system without intermediate losses the performance everything goes smoothly with quite good results. So, here there are my questions regarding HierLID_utt:

· **Input Data.** I am preparing the input data for the auxiliary task _lid_utt_ as follows:

    - batch['lid_utt'] # -- (batch, 1,) since it is a single token representing the language of the token.
    - batch['lid_utt_lengths'] # -- (batch,) representing the lenght of each _lid_utt_ input sample: always 1

Is this the correct way to prepare the input data for the auxiliary task?

· My second question:is the _linear conditioning layer_ used in the the 3rd interloss (language identification) the same as the one used in the rest of the intermediate losses and final prediction? Or did you used a different layer, i.e., you did not shared the parameters? According to the way I prepare the data it seems I am trying to force the model to align the latent representation of the 3rd layer to one single token, which I am not sure is helping to model speech later.

I read the paper but I am still have doubts. Thanks in advance, David.

**Comment by wanchichen** [comment]
Your data setup looks correct to me. I have only tried auto regressive decoders with the intermediate loss, so I’m not sure how well the technique will extend to MaskCTC.

We share the linear projection parameters across all of the intermediate predictions layers **and** the final CTC layer. 

Two questions: 

Are you down weighting the intermediate loss at all? 

Does the entire network fail to learn, or only the intermediate loss, or the final loss?

**Comment by david-gimeno** [comment]
Thanks for your prompt reply! Regarding your questions:

`1) Are you down weighting the intermediate loss at all?`

My settings are as follows: _ctc_weight=0.3_, and _interctc_weight=0.5_

`2) Does the entire network fail to learn, or only the intermediate loss, or the final loss?`

I am not sure if I understand the question. Let's compare these details. After 40k training steps, these are the results for:

· **Vanilla Conformer:** TRAIN LOSS=66.61683533063042 || VAL LOSS=28.324 || VAL CER=5.625%
· **HierLIDutt Conformer:** TRAIN LOSS=96.11445122377435 || VAL LOSS=98.687 | VAL CER=100.0%

Perhaps, the HierLIDutt approach needs more steps to converge but the performance gap is quite big. Another option could be that I am trying with a non-autoregressive framework, such as MaskCTC but I remember that I trained a only-CTC model, i.e., setting _ctc_weight=1.0_ and the results were quite similar. What's your opinion?

**Comment by wanchichen** [comment]
Hmm those settings should be fine, although the interctc weight could be lower. 

Are you logging the value of the intermediate loss? Generally that should converge extremely quickly. 
Have you also tried a normal self-conditioned CTC model instead? You probably only need to train for a few steps to see if it converges compared to the hierarchical model.

**Comment by david-gimeno** [comment]
Following your suggestion, I saw that the normal self-conditioned CTC model was working as expected. Subsequently, after reviewing the code and every single detail carefully, I found that was a problem with the tokenizer, which was adding a space token! Now, the model seems to learn properly.

Thanks. I really appreciate the time you invested in replying to me. I would like to apologize for the inconvenience because, as usually happens in the world of programming, the bug was a small, little detail in the code :)
